module/zfs/zfeature_common.c

```diff
@@ -242,4 +242,15 @@ zpool_feature_init(void)
 	    "Support for blocks larger than 128KB.", B_FALSE, B_FALSE, B_FALSE,
 	    large_blocks_deps);
 	}
+
+	{
+	static const spa_feature_t large_dnode_deps[] = {
+		SPA_FEATURE_EXTENSIBLE_DATASET,
+		SPA_FEATURE_NONE
+	};
+	zfeature_register(SPA_FEATURE_LARGE_DNODE,
```

Comment:
ZoL should sync up with https://github.com/illumos/illumos-gate/commit/ca0cc3918a1789fa839194af2a9245f801a06b1a (5959 clean up per-dataset feature count code), and then you should make this a ZFEATURE_FLAG_PER_DATASET.  That should clean up a bunch of your feature flag handling code.